mlflow_faculty/filter.py

```diff
@@ -0,0 +1,402 @@
+    PARAM = "parameter"
+    METRIC = "metric"
+    TAG = "tag"
+
+
+ATTRIBUTE_IDENTIFIERS = {"attribute", "attributes", "attr", "run"}
+RUN_ID_IDENTIFIERS = {"id", "run_id"}
+PARAM_IDENTIFIERS = {"param", "params", "parameter", "parameters"}
+METRIC_IDENTIFIERS = {"metric", "metrics"}
+TAG_IDENTIFIERS = {"tag", "tags"}
+
+COMPARISON_OPERATOR_MAPPING = {
+    "=": ComparisonOperator.EQUAL_TO,
+    "!=": ComparisonOperator.NOT_EQUAL_TO,
+    ">": ComparisonOperator.GREATER_THAN,
+    ">=": ComparisonOperator.GREATER_THAN_OR_EQUAL_TO,
+    "<": ComparisonOperator.LESS_THAN,
+    "<=": ComparisonOperator.LESS_THAN_OR_EQUAL_TO,
+}
+DISCRETE_KEY_TYPES = {_KeyType.RUN_ID, _KeyType.STATUS, _KeyType.TAG}
+DISCRETE_OPERATORS = {
+    ComparisonOperator.DEFINED,
+    ComparisonOperator.EQUAL_TO,
+    ComparisonOperator.NOT_EQUAL_TO,
+}
+
+
+class MatchesNothing(Exception):
+    """Raised when a filter always matches nothing."""
+
+    pass
+
+
+def build_search_runs_filter(experiment_ids, filter_string, view_type):
+    """Build a filter from the inputs to search_runs in the tracking store."""
+
+    filter_parts = []
+
+    if experiment_ids is not None:
+        filter_parts.append(filter_by_experiment_id(experiment_ids))
+
+    deleted_at_filter = filter_by_mlflow_view_type(view_type)
+    if deleted_at_filter is not None:
+        filter_parts.append(deleted_at_filter)
+
+    if filter_string is not None and filter_string.strip() != "":
+        filter_parts.append(parse_filter_string(filter_string))
+
+    if len(filter_parts) == 0:
+        return None
+    elif len(filter_parts) == 1:
+        return filter_parts[0]
+    else:
+        return CompoundFilter(LogicalOperator.AND, filter_parts)
+
+
+def filter_by_experiment_id(experiment_ids):
+    """Build a filter that a run is in one of a sequence of experiment IDs."""
+
+    if len(experiment_ids) == 0:
+        # Cannot build a filter for this
+        raise MatchesNothing()
+
+    parts = [
+        ExperimentIdFilter(ComparisonOperator.EQUAL_TO, int(experiment_id))
+        for experiment_id in experiment_ids
+    ]
+
+    if len(parts) == 1:
+        return parts[0]
+    else:
+        return CompoundFilter(LogicalOperator.OR, parts)
+
+
+def filter_by_mlflow_view_type(view_type):
+    """Build a filter for an MLflow view type.
+
+    Parameters
+    ----------
+    view_type : mlflow.entities.ViewType
+
+    Returns
+    -------
+    A Faculty library filter object, or None if no filter is to be applied.
+    """
+    if view_type == ViewType.ACTIVE_ONLY:
+        return DeletedAtFilter(ComparisonOperator.DEFINED, False)
+    elif view_type == ViewType.DELETED_ONLY:
+        return DeletedAtFilter(ComparisonOperator.DEFINED, True)
+    elif view_type == ViewType.ALL:
+        return None
+    else:
+        raise ValueError("Invalid ViewType: {}".format(view_type))
+
+
+def parse_filter_string(mlflow_filter_string):
+    """Parse an MLflow filter string into a Faculty filter object."""
```

Comment:
Missing full docstring?

Reply:
Since this is only really used internally, I don't think a full docstring is really necessary here.

Reply:
In that case, does it make sense to make it private?

Reply:
Fair, will do.